# Functional_region_annotation_pipeline
## Overview
An extension of existing variant annotation pipeline which adds extra functional scores and region-based annotations. The objective of this improved variant annotation pipeline is to gather as much information about the variant input as possible by pure bioinformatic methods. As a result, more informed and educated interpretations and decisions on genome variants can be made before moving onto the subsequent variant validation experiments and *in vitro/in vivo* functional studies of the variants.
## Usage
### 1. Preparing the input_files.txt
### 2. Setting up the snake_conf.yaml
### 3. Running the pipeline via snakeamke commands
## Understanding the final output
